### PR TITLE
Keyring support

### PR DESCRIPTION
**Pull request by vit-tomica**
### SAF keyring configuration for Zowe
This PR enhances certificate configuration scripts to store certificates in keyrings.
The idea is to provide analogous actions for keyrings to what we currently have for keystores, i.e. generating a local zowe CA and zowe cert, importing external certs, delete certs or generally doing certs cleanup 

#### Motivation:
- run Zowe on keyrings
- when configuring keyrings, do as much automation as possible 
- if possible avoid writing a lot of documenation (commands for each ESMs etc...)

How much java keytool can handle SAF keyrings
it can: 
- generate and connect a certificate with a keyring, creates the keyring if does not exist.
- import/export a certificate from/to a keystore, creates the keyring if does not exist.
- disconnect a certifice from a keyring.
- there is also a java hwkeytool utility that can take advantage of crypto hardware.

it can't:
- delete keyring
- delete a certificate from RACF database.

The **keyring-util program** (leverages R_datalib API) provides keyring actions that the Java keytool can't do.
it can: 
- delete keyring
- delete a certificates from RACF database
- refresh DIGTCERT class

The keyring-util's functions could be useful in scenarios like:
- cleanup keyring configuration
- renewing certs
- automation
- if keyring configuration fails in the middle of the process, we can cleanup partial setup and start from scratch
______________________________________________________________________________ 

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Necessary documentation (if appropriate) have been added / updated
- [ ] DCO signoffs have been added to all commits, including this PR

#### PR type
What type of changes does your PR introduce to Zowe? _Put an `x` in the box that applies to this PR. If you're unsure about any of them, don't hesitate to ask._ 
- [ ] Bugfix 
- [ ] Feature 
- [ ] Other... Please describe:

#### Relevant issues

Fixes 

#### Changes proposed in this PR

- 

#### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

#### Does this PR do something the person installing Zowe should know about?

---
* Affected function: _general area of interest_ *

---
* Description: _1 line description_ *

---
* Part: _name of customizable file involved_  *

---
_multi-line description_

#### Is there a related doc issue or Pull Request? 

Doc issue/PR number: 
https://github.com/zowe/docs-site/pull/1214

### CODE REVIEW

**Comment by OnnoVdT** [comment]
assigned myself to have a closer look when I have some time as this is interesting stuff

**Comment by vit-tomica** [comment]
Thanks Onno.
From user perspective, the action should be simple - setting a keyring name for the ZOWE_KEYRING= variable  in the zowe-setup-certificates.env, the rest should be handled by the zowe-setup-certificates.sh script

**Comment by OnnoVdT** [comment]
Do ACF2 and TSS support the R_datalib API ? I thought this was RACF only (but could be mistaken).
Also, we heard that they current script for keystores is extremely slow on ZD&T (takes about an hour),. This is likely because we drive several invocations of the keytool utility, which in turn drives Java.
We are using keytool here as well, so the keyring support might end up being slow as well. We should try this out and maybe consider to see if something can be doen with direct RACF/ACF2/TSS commands. That would be another issue/pull request, just bringing it to your attention here.

**Comment by vit-tomica** [comment]
Yes, they do. The Java keytool leverages R_datalib API and we recently used this in Zowe to read zosmf keyring to get and trust the zosmf cert. However, I have found a bug in TSS/ACF2 when creating/deleting keyring using R_datalib. Security team is working on the fix and this PR will not work for TSS/ACF2 until it's fixed.

You're right about the ZD&T. It'll be slow because of the keytool commands. In this case, usual security commands make sense.

**Comment by 1000TurquoisePogs** [comment]
@timgerstel @vit-tomica Tim's working on using the JWT key info to enable SSO on our other servers. Can you two please collaborate on if these changes will work with ongoing SSO efforts? We don't want the way SAF keyring support is done to hold back SSO work, so just double-check on the files & formats.

**Comment by timgerstel** [comment]
@vit-tomica @1000TurquoisePogs  I do not believe this pull request will interfere with the work to enable SSO.  All necessary files will still exist and are in the correct format.

**Comment by plavjanik** [comment]
This looks great. What privileges does a user need to have to be able to run it? Is there a related documentation PR?

**Comment by vit-tomica** [comment]
@plavjanik I don't know exactly what all the privileges are but READ access to all of the IRR.DIGTCERT.* resources is enough + UPDATE access to IRR.DIGTCERT.CONNECT is needed to set up trust with z/OSMF. The READ access to all IRR.DIGTCERT.* resources is a bit excessive but it generally allows to perform actions only on certificates and keyrings owned by the ACID that has the READ access

There is no doc yet.

I'm also working on JCL configuration. I plan to remove the part that compile/link the keyring-util program and I'll replace it with the prebuilt program instead.

**Comment by vit-tomica** [comment]
The [ACF2 PTF](https://support.broadcom.com/download-center/solution-detail.html?aparNo=SO11887&os=z%2FOS) has been published. It fixes the bug that prevents keytool from adding/importing a certificate to the keyring. A Top Secret fix should be published soon as well.

**Comment by vit-tomica** [comment]
Related doc PR - https://github.com/zowe/docs-site/pull/1214

File: .pax/prepare-workspace.sh [hunk 1]
```diff
@@ -23,6 +23,7 @@ set -x
 
 # expected workspace layout:
```
Comment (OnnoVdT) [inline .pax/prepare-workspace.sh:24]: had to think again what and where this was. Can you update this line to say the following and remove all doubt
`# expected input workspace layout ($ROOT_DIR):`

File: .pax/prepare-workspace.sh [hunk 2]
```diff
@@ -23,6 +23,7 @@ set -x
...
 # ./.pax/mediation/
+# ./.pax/keyring-util/
```
Comment (OnnoVdT) [inline .pax/prepare-workspace.sh:26]: keep 'm in alphabetical order so it lines up with output of ls

File: .pax/prepare-workspace.sh [hunk 3]
```diff
@@ -199,5 +209,5 @@ echo "[$SCRIPT_NAME] done"
 # ${PAX_WORKSPACE_DIR}/ascii/zowe-${ZOWE_VERSION}/
 # ${PAX_WORKSPACE_DIR}/content/zowe-${ZOWE_VERSION}/
 # ${PAX_WORKSPACE_DIR}/mediation/  # already present
-
+# ${PAX_WORKSPACE_DIR}/keyring-util/  # already present
```
Comment (OnnoVdT) [inline .pax/prepare-workspace.sh:212]: keep 'm in alphabetical order

File: files/jcl/ZWEKRING.jcl [hunk 4]
```diff
@@ -0,0 +1,376 @@
+//ZWEKRING JOB
+//*
+//* This program and the accompanying materials are made available
+//* under the terms of the Eclipse Public License v2.0 which
+//* accompanies this distribution, and is available at
+//* https://www.eclipse.org/legal/epl-v20.html
+//*
+//* SPDX-License-Identifier: EPL-2.0
+//*
+//* Copyright Contributors to the Zowe Project. 2018, 2020
```
[the remark this answers is not in the document]
Comment (vit-tomica) [reply files/jcl/ZWEKRING.jcl]: Done.

File: files/jcl/ZWEKRING.jcl [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+//*    meet your system requirements.
+//*
+//* 2) Update the SET PRODUCT= statement to match your security
+//*    product.
+//*
+//* 3) Update the SET ZOWEUSER= statement to match the desired
+//*    user ID for the ZOWE started task.
+//*
+//* 4) Update the SET ZOWERING= statement to match the desired
+//*    name of the keyring owned by the ZOWEUSER.
```
[the remark this answers is not in the document]
Comment (vit-tomica) [reply files/jcl/ZWEKRING.jcl]: Done.

File: files/jcl/ZWEKRING.jcl [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+//*
+//* 5) Update the SET LABEL= statement with the name of the Zowe
+//*    certificate that will be added to the RACF database or that
```
[the remark this answers is not in the document]
Comment (vit-tomica) [reply files/jcl/ZWEKRING.jcl]: Done.

File: files/jcl/ZWEKRING.jcl [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+//* Before using this JCL, you will have to make the following
+//* modifications:
+//*
+//* 1) Add job name and job parameters to the JOB statement, to
+//*    meet your system requirements.
+//*
+//* 2) Update the SET PRODUCT= statement to match your security
+//*    product.
+//*
+//* 3) Update the SET ZOWEUSER= statement to match the desired
```
[the remark this answers is not in the document]
Comment (vit-tomica) [reply files/jcl/ZWEKRING.jcl]: Done.

File: files/jcl/ZWEKRING.jcl [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+//*    user ID for the ZOWE started task.
```
[the remark this answers is not in the document]
Comment (vit-tomica) [reply files/jcl/ZWEKRING.jcl]: Done.

File: .pax/prepare-workspace.sh (hunk 1 above)
Comment (vit-tomica) [reply .pax/prepare-workspace.sh]: Done.

File: .pax/prepare-workspace.sh (hunk 2 above)
Comment (vit-tomica) [reply .pax/prepare-workspace.sh:25]: Done.

File: .pax/prepare-workspace.sh (hunk 3 above)
Comment (vit-tomica) [reply .pax/prepare-workspace.sh:211]: Done.

**Comment by vit-tomica** [comment]
@OnnoVdT Hi Onno, just a reminder that all your comments have been addressed.

**Comment by OnnoVdT** [comment]
sorry, I saw you did but forgot to approve. That is corrected now

**Pull request merged**